File: eRents.Domain/Models/Address.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using Microsoft.EntityFrameworkCore;

namespace eRents.Domain.Models;

/// <summary>
/// Address Value Object - represents a physical address as a cohesive unit
/// Uses EF Core's [Owned] attribute to embed in entities without separate table
/// </summary>
[Owned]
public class Address
{
    [StringLength(255)]
    public string? StreetLine1 { get; set; }
    
    [StringLength(255)]
    public string? StreetLine2 { get; set; }
    
    [StringLength(100)]
    public string? City { get; set; }
    
    [StringLength(100)]
    public string? State { get; set; }
    
    [StringLength(100)]
    public string? Country { get; set; }
    
    [StringLength(20)]
    public string? PostalCode { get; set; }
    
    [Column(TypeName = "decimal(9, 6)")]
    public decimal? Latitude { get; set; }
    
    [Column(TypeName = "decimal(9, 6)")]
    public decimal? Longitude { get; set; }
    
    /// <summary>
    /// Gets the full address as a formatted string
    /// </summary>
    public string GetFullAddress()
    {
        var parts = new[] { StreetLine1, StreetLine2, City, State, Country, PostalCode }
            .Where(x => !string.IsNullOrEmpty(x));
        return string.Join(", ", parts);
    }
    
    /// <summary>
    /// Gets the basic street address (street lines only)
    /// </summary>
    public string GetStreetAddress()
    {
        var parts = new[] { StreetLine1, StreetLine2 }
            .Where(x => !string.IsNullOrEmpty(x));
        return string.Join(", ", parts);
    }
    
    /// <summary>
    /// Gets the location part of the address (city, state, country)
    /// </summary>
    public string GetLocationString()
    {
        var parts = new[] { City, State, Country }
            .Where(x => !string.IsNullOrEmpty(x));
        return string.Join(", ", parts);
    }
    
    /// <summary>
    /// Determines if the address is considered empty (no essential information)
    /// </summary>
    public bool IsEmpty => string.IsNullOrEmpty(StreetLine1) && string.IsNullOrEmpty(City);
    
    /// <summary>
    /// Determines if the address has coordinates
    /// </summary>
    public bool HasCoordinates => Latitude.HasValue && Longitude.HasValue;
    
    /// <summary>
    /// Creates an Address from separate components
    /// </summary>
    public static Address Create(
        string? streetLine1 = null,
        string? streetLine2 = null,
        string? city = null,
        string? state = null,
        string? country = null,
        string? postalCode = null,
        decimal? latitude = null,
        decimal? longitude = null)
    {
        return new Address
        {
            StreetLine1 = streetLine1?.Trim(),
            StreetLine2 = streetLine2?.Trim(),
            City = city?.Trim(),
            State = state?.Trim(),
            Country = country?.Trim(),
            PostalCode = postalCode?.Trim(),
            Latitude = latitude,
            Longitude = longitude
        };
    }
    
    /// <summary>
    /// Value object equality comparison
    /// </summary>
    public override bool Equals(object? obj)
    {
        if (obj is not Address other) return false;
        
        return StreetLine1 == other.StreetLine1 &&
               StreetLine2 == other.StreetLine2 &&
               City == other.City &&
               State == other.State &&
               Country == other.Country &&
               PostalCode == other.PostalCode &&
               Latitude == other.Latitude &&
               Longitude == other.Longitude;
    }
    
    /// <summary>
    /// Value object hash code
    /// </summary>
    public override int GetHashCode()
    {
        return HashCode.Combine(
            StreetLine1,
            StreetLine2, 
            City,
            State,
            Country,
            PostalCode,
            Latitude,
            Longitude);
    }
    
    public override string ToString() => GetFullAddress();
} 

File: e_rents_desktop/lib/models/address.dart

class Address {
  final String? streetLine1;
  final String? streetLine2;
  final String? city;
  final String? state;
  final String? country;
  final String? postalCode;
  final double? latitude;
  final double? longitude;

  const Address({
    this.streetLine1,
    this.streetLine2,
    this.city,
    this.state,
    this.country,
    this.postalCode,
    this.latitude,
    this.longitude,
  });

  factory Address.fromJson(Map<String, dynamic> json) {
    double? _toDouble(dynamic v) {
      if (v == null) return null;
      if (v is num) return v.toDouble();
      if (v is String) return double.tryParse(v);
      return null;
    }
    // Backend AddressResponse uses 'street', frontend uses 'streetLine1'
    final streetLine1 = json['streetLine1'] as String? ?? json['street'] as String?;
    return Address(
      streetLine1: streetLine1,
      streetLine2: json['streetLine2'] as String?,
      city: json['city'] as String?,
      state: json['state'] as String?,
      country: json['country'] as String?,
      postalCode: json['postalCode'] as String? ?? json['zipCode'] as String?,
      latitude: _toDouble(json['latitude']),
      longitude: _toDouble(json['longitude']),
    );
  }

  Map<String, dynamic> toJson() => <String, dynamic>{
        'streetLine1': streetLine1,
        'streetLine2': streetLine2,
        'city': city,
        'state': state,
        'country': country,
        'postalCode': postalCode,
        'latitude': latitude,
        'longitude': longitude,
      };

  factory Address.empty() {
    return const Address(
      streetLine1: '',
      streetLine2: '',
      city: '',
      state: '',
      country: '',
      postalCode: '',
      latitude: null,
      longitude: null,
    );
  }

  String getFullAddress() {
    final parts = [
      streetLine1,
      streetLine2,
      city,
      state,
      country,
      postalCode,
    ].where((part) => part?.isNotEmpty == true);
    return parts.join(', ');
  }

  String getStreetAddress() {
    final parts = [
      streetLine1,
      streetLine2,
    ].where((part) => part?.isNotEmpty == true);
    return parts.join(', ');
  }

  String getCityStateCountry() {
    final parts = [
      city,
      state,
      country,
    ].where((part) => part?.isNotEmpty == true);
    return parts.join(', ');
  }

  bool get isEmpty => (streetLine1?.isEmpty ?? true) && (city?.isEmpty ?? true);
  bool get isNotEmpty => !isEmpty;

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;
    if (other is! Address) return false;
    return streetLine1 == other.streetLine1 &&
        streetLine2 == other.streetLine2 &&
        city == other.city &&
        state == other.state &&
        country == other.country &&
        postalCode == other.postalCode &&
        latitude == other.latitude &&
        longitude == other.longitude;
  }

  @override
  int get hashCode => Object.hash(
        streetLine1,
        streetLine2,
        city,
        state,
        country,
        postalCode,
        latitude,
        longitude,
      );

  @override
  String toString() => getFullAddress();
}